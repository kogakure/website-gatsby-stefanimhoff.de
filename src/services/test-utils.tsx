import React from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';

import { EmojifyProvider, LocalizedDateProvider } from '../contexts';

const customRender = (
  node: React.ReactElement,
  options?: RenderOptions
): RenderResult => {
  const rendered = render(
    <EmojifyProvider>
      <LocalizedDateProvider>{node}</LocalizedDateProvider>
    </EmojifyProvider>,
    options
  );

  return {
    ...rendered,
    rerender: (ui: React.ReactElement, rerenderOptions?: RenderOptions) =>
      customRender(ui, { container: rendered.container, ...options }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
