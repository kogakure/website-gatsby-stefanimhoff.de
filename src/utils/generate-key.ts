export const generateKey = (prefix: string) => `${prefix}-${Date.now()}`;
