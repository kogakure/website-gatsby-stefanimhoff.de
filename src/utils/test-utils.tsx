import React from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { EmojifyProvider, LocalizedDateProvider } from '../contexts';
import { ThemeType } from '../typings/Theme';
import { lightTheme } from '../themes';

interface CustomTheme {
  theme: ThemeType;
}

type RenderOptionsWithTheme = RenderOptions & CustomTheme;

const customRender = (
  node: React.ReactElement,
  options: RenderOptionsWithTheme = { theme: lightTheme }
): RenderResult => {
  const { theme } = options;
  const rendered = render(
    <ThemeProvider theme={theme}>
      <EmojifyProvider>
        <LocalizedDateProvider>{node}</LocalizedDateProvider>
      </EmojifyProvider>
    </ThemeProvider>,
    options
  );

  return {
    ...rendered,
    rerender: (
      ui: React.ReactElement,
      rerenderOptions?: RenderOptionsWithTheme
    ) => customRender(ui, { container: rendered.container, ...options }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
