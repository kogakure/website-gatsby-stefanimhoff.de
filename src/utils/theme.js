/* eslint-disable */
(function() {
  const root = document.getElementsByTagName('html')[0];
  window.__onThemeChange = function() {};

  function setTheme(newTheme) {
    window.__theme = newTheme;
    preferredTheme = newTheme;
    currentTheme = newTheme === 'light' ? 'dark' : 'light';
    root.classList.add(newTheme);
    root.classList.remove(currentTheme);
    window.__onThemeChange(newTheme);
  }

  let preferredTheme;

  try {
    preferredTheme = localStorage.getItem('theme');
  } catch (err) {}

  window.__setPreferredTheme = function(newTheme) {
    setTheme(newTheme);
    try {
      localStorage.setItem('theme', newTheme);
    } catch (err) {}
  };

  const darkQuery = window.matchMedia('(prefers-color-scheme: dark)');

  darkQuery.addListener(function(e) {
    window.__setPreferredTheme(e.matches ? 'dark' : 'light');
  });

  setTheme(preferredTheme || (darkQuery.matches ? 'dark' : 'light'));
})();
