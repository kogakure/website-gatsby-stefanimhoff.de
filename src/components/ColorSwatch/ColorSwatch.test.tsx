import React from 'react';
import { axe } from 'jest-axe';
import user from '@testing-library/user-event';

import { render, fireEvent } from '../../utils/test-utils';

import { ColorSwatch } from '.';

describe('ColorSwatch', () => {
  const color = '#005CAF';
  const rgbColor = 'rgb(0, 92, 175)';
  const hslColor = 'hsl(208, 100%, 34%)';
  const name = 'Lapis Lazuli';
  const description =
    'The soft, slightly purplish blue associated with the semi-precious stone.';

  test('renders correctly', () => {
    const { container } = render(
      <ColorSwatch color={color} name={name} description={description} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly with name', () => {
    const { getByText } = render(<ColorSwatch color={color} name={name} />);

    expect(getByText(/lapis lazuli/i)).toBeInTheDocument();
  });

  test('renders correctly with description', () => {
    const { getByText } = render(
      <ColorSwatch color={color} description={description} />
    );

    expect(getByText(/purplish blue/i)).toBeInTheDocument();
  });

  test('renders correctly with RGB value', () => {
    const { getByText } = render(<ColorSwatch color={rgbColor} />);

    expect(getByText(/#005CA/)).toBeInTheDocument();
  });

  test('renders correctly with HSL value', () => {
    const { getByText } = render(<ColorSwatch color={hslColor} />);

    expect(getByText(/#005CA/)).toBeInTheDocument();
  });

  test('show color information on click', async () => {
    const { getByText, findByLabelText } = render(
      <ColorSwatch color={color} name={name} description={description} />
    );

    user.click(await findByLabelText('Information'));
    expect(getByText(/#005CA/i)).toBeInTheDocument();

    user.click(await findByLabelText('Information'));
    expect(getByText(/lapis lazuli/i)).toBeInTheDocument();
  });

  test('should cycle through information icon', async () => {
    const { findByLabelText } = render(
      <ColorSwatch color={color} name={name} description={description} />
    );

    expect(document.body).toHaveFocus();
    user.tab();
    expect(await findByLabelText('Information')).toHaveFocus();
  });

  test('show color information on Enter keyUp', async () => {
    const { getByText, findByLabelText } = render(
      <ColorSwatch color={color} name={name} description={description} />
    );

    fireEvent.keyUp(await findByLabelText('Information'), {
      key: 'Enter',
      code: 13,
    });
    expect(getByText(/#005CA/i)).toBeInTheDocument();

    fireEvent.keyUp(await findByLabelText('Information'), {
      key: 'Enter',
      code: 13,
    });
    expect(getByText(/lapis lazuli/i)).toBeInTheDocument();
  });

  test('is accessible', async () => {
    const { container } = render(
      <ColorSwatch color={color} name={name} description={description} />
    );
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
