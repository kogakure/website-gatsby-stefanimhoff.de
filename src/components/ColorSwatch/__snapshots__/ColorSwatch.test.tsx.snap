// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorSwatch renders correctly 1`] = `
.c0 {
  border: 1px solid var(--colorBorderSoft);
  border-radius: 5px;
  box-shadow: 0 5px 50px rgba(0,0,0,0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 0.7rem;
  width: 12rem;
}

.c1 {
  background: lightgrey;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  height: 6rem;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 30px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  right: 0;
  top: 0;
  width: 30px;
}

.c4 {
  fill: var(--colorTextFullContrast);
  font-size: 1rem;
}

.c2 {
  background: var(--colorBackgroundFullContrast);
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  color: var(--colorTextFullContrast);
  min-height: 7rem;
  padding: 0.8rem 0.6rem;
  position: relative;
}

.c5 {
  font-size: 0.8rem;
  font-weight: bold;
  line-height: 1;
  padding-bottom: 0.3rem;
  padding-right: 20px;
}

.c6 {
  margin-bottom: 0.5rem;
  padding-right: 20px;
}

<div
  class="c0"
>
  <div
    class="c1"
    style="background-color: rgb(0, 92, 175);"
  />
  <div
    class="c2"
  >
    <div
      aria-label="Information"
      class="c3"
      role="button"
      tabindex="0"
    >
      <svg
        aria-label="Information Icon"
        class="c4"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm0-2a8 8 0 100-16 8 8 0 000 16zM11 7h2v2h-2V7zm0 4h2v6h-2v-6z"
        />
      </svg>
    </div>
    <div
      class="c5"
    >
      Lapis Lazuli
    </div>
    <div
      class="c6"
    >
      The soft, slightly purplish blue associated with the semi-precious stone.
    </div>
  </div>
</div>
`;
